lib/Dialect/Arc/Transforms/LatencyRetiming.cpp

```diff
@@ -55,61 +56,123 @@ struct LatencyRetimingPattern : OpRewritePattern<StateOp> {
-    auto predState = input.getDefiningOp<StateOp>();
-    if (!predState)
+  assert(isa<mlir::CallOpInterface>(op.getOperation()) &&
+         "state and call operations call arcs and thus have to implement the "
+         "CallOpInterface");
+  auto callOp = cast<mlir::CallOpInterface>(op.getOperation());
+
+  for (auto input : callOp.getArgOperands()) {
+    auto predOp = input.getDefiningOp<ClockedOpInterface>();
+
+    // Only support call and state ops for the predecessors as well.
+    if (!predOp || !isa<CallOp, StateOp>(predOp.getOperation()))
       return failure();
 
-    if (predState->hasAttr("name") || predState->hasAttr("names"))
+    // Conditions for both StateOp and CallOp
+    if (predOp->hasAttr("name") || predOp->hasAttr("names"))
       return failure();
 
-    if (predState == op)
+    // Check for a use-def cycle since we can be in a graph region.
+    if (predOp == op)
       return failure();
 
-    if (predState.getLatency() != 0 && op.getLatency() != 0 &&
-        predState.getClock() != op.getClock())
+    if (predOp.getClock() && op.getClock() &&
+        predOp.getClock() != op.getClock())
       return failure();
 
-    if (predState.getEnable() || predState.getReset())
+    if (predOp->getParentRegion() != op->getParentRegion())
       return failure();
 
-    if (llvm::any_of(predState->getUsers(),
+    if (hasEnableOrReset(predOp))
+      return failure();
+
+    // Check that the predecessor state does not have another user since then
+    // we cannot change its latency attribute without also changing it for the
+    // other users. This is not supported yet and thus we just fail.
+    if (llvm::any_of(predOp->getUsers(),
                      [&](auto *user) { return user != op; }))
       return failure();
 
-    predecessors.insert(predState);
-    minPrevLatency = std::min(minPrevLatency, predState.getLatency());
+    // We check that all clocks are the same if present. Here we remember that
+    // clock. If none of the involved operations have a clock, they must have
+    // latency 0 and thus `minPrevLatency = 0` leading to early failure below.
+    if (!clock) {
+      if (predOp.getClock())
+        clock = predOp.getClock();
+      if (auto clockDomain = predOp->getParentOfType<ClockDomainOp>())
+        clock = clockDomain.getClock();
+    }
+
+    predecessors.insert(predOp);
+    minPrevLatency = std::min(minPrevLatency, predOp.getLatency());
   }
 
   if (minPrevLatency == 0 || minPrevLatency == UINT_MAX)
     return failure();
 
-  op.setLatency(op.getLatency() + minPrevLatency);
-  for (auto prevStateOp : predecessors) {
-    if (!op.getClock() && !op->getParentOfType<ClockDomainOp>())
-      op.getClockMutable().assign(prevStateOp.getClock());
-
-    statistics.latencyUnitsSaved += minPrevLatency;
-    auto newLatency = prevStateOp.getLatency() - minPrevLatency;
-    prevStateOp.setLatency(newLatency);
+  auto setLatency = [&](Operation *op, uint64_t newLatency, Value clock) {
```

Comment:
How come this is a lambda? (Assuming it's just a formatting/simplicity point but curious in case there's some deeper reason I missed)

Reply:
Two places call this logic, i.e., code reuse. It's a lambda rather than a regular function because that allows capturing, e.g., `symCache` rather than adding it as an additional func argument.

Reply:
Ah cool, I overlooked the capturing!